### ISSUE TITLE
ADFS Error with SMS/Email Challenge

### ISSUE THREAD

**Issue by dwright726** [sbidy/privacyIDEA-ADFSProvider#22]
I am having an issue with SMS/Email challenges when using this provider, the TOTP OTP codes seem to work fine.  When logging in ADFS is logging the following error:

triggerChallenge: The remote server returned an error: (400) Bad Request.

System.Net.WebException: The remote server returned an error: (400) Bad Request.
   at System.Net.WebClient.UploadValues(Uri address, String method, NameValueCollection data)
   at privacyIDEAADFSProvider.OTPprovider.triggerChallenge(String OTPuser, String realm, String token)

I have configured the config.xml file with an admin account and have tired this several different ways however, with the various configurations I have tried I receive additional errors about the admin username/password which makes me believe that the current configuration is the correct one.

My question at this point is what am I doing wrong?  What else do I need to try?

**Comment by sbidy** [comment]
Hey,

What privacyIDEA version you are using?

**Comment by dwright726** [comment]
Version 2.23.5

**Comment by dwright726** [comment]
Side Note, upon realizing I was on 2.23.5 I went ahead and upgraded Pidea to 3.0.2 (Ubuntu 16.04 LTS) and I am still receiving the same error?

**Comment by dwright726** [comment]
So after some additional troubleshooting I have determined that issue was with my PI configuration.  (Really wish there was an easy to use Test SMS button to verify SMS Provider Configuration without being required to enter a PIN or that PI documentation had better information on testing SMS configurations, also wish API documentation included some usage examples instead of just listing all the possible options)  After resolving issues with the SMS gateway configuration (using an HTTP based SMS provider MessageMedia) and finally verifying that SMS was being triggered via the challenge response the problem ended up being that my test account was falling under my admin realm resolver instead of my normal user resolver, I am not 100% sure why however, I will say that the Admin resolver was set as the default resolver and once I changed the default resolver to the normal user resolver everything started working as expected.  I then created a policy within PI to send the SMS/Email OTP during authentication and applied this to all realms.  

I am now seeing the results I was looking for, when authenticating against ADFS a SMS message and an Email message containing a OTP is sent to the user during authentication and those codes all work when ADFS prompts for a OTP (HOTP/TOTP tokens still work as well).  For my use case this provides the most flexibility to users attempting to access the ADFS authenticated resources.
